File: leak-examiner/scan.py
import re
import json
import sys
import requests
import re

def extract_comments(content, url=""):
    html_comments = {}
    js_comments = {}

    # Extract HTML comments
    html_pattern = r"<!--(.*?)-->"
    for match in re.finditer(html_pattern, content, re.DOTALL):
        comment = match.group(1).strip()
        html_comments[match.start() + 1] = f"{url} {comment}" if url else comment

    # Extract JavaScript comments
    js_pattern = r"//(.*?)$|/\*(.*?)\*/"
    for script_match in re.finditer(r"<script\b[^>]*>(.*?)</script>", content, re.DOTALL | re.IGNORECASE):
        script_comments = {}
        script_content = script_match.group(1)
        for line_number, line in enumerate(script_content.splitlines(), start=1):
            for match in re.finditer(js_pattern, line):
                comment = match.group(1) or match.group(2)
                script_comments[line_number] = comment.strip()
        js_comments[script_match.start() + len("<script>")] = script_comments

    return {"html": html_comments, "js": js_comments}


def scan_for_prohibited(comments_html, comments_js):
    prohibited_findings = {}
    try:
        with open("prohib.json", "r") as f:
            prohibited_words = json.load(f)
    except FileNotFoundError:
        print("Error: Could not find prohib.json file")
        return prohibited_findings

    for value in prohibited_words:
        for line_number, comment in comments_html.items():
            if value.lower() in comment.lower():
                prohibited_findings[line_number] = f"PROHIBITED VALUE ({value}) in comment {comment}"

        for script_location, script_comments in comments_js.items():
            for line_number, comment in script_comments.items():
                if value.lower() in comment.lower():
                    # Adjust line number to reflect location within script block
                    prohibited_findings[script_location + line_number - 1] = f"PROHIBITED VALUE ({value}) in comment {comment}"

    return prohibited_findings

def main():
    if len(sys.argv) < 2:
        print("Usage: scan.py <url>")
        exit(1)

    url = sys.argv[1]
    try:
        content = requests.get(url).text
    except requests.exceptions.RequestException as e:
        print(f"Error: Could not retrieve content from URL: {e}")
        exit(1)

    extracted_comments = extract_comments(content, url)
    html_comments = extracted_comments["html"]
    js_comments = extracted_comments["js"]

    if html_comments:
        print("\nComments discovered (html):")
        for line_number, comment in html_comments.items():
            print(f"[{line_number}] {comment}")

    if js_comments:
        print("\nComments discovered (JS):")
        for script_location, script_comments in js_comments.items():
            for line_number, comment in script_comments.items():
                print(f"[Script location: {script_location + 1}, Line: {line_number}] {comment}")

    prohibited_findings = scan_for_prohibited(html_comments, js_comments)
    if prohibited_findings:
        print("\n**************** PROHIBITED COMMENTS DISCOVERED ****************")
        for line_number, message in prohibited_findings.items():
            print(message)

if __name__ == "__main__":
    import requests  # You may need to install the requests library
    main()

File: secrets/scan.java
import java.io.IOException;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CommentScanner {

    public static Map<Integer, String> extractHtmlComments(String content, String url) {
        Map<Integer, String> htmlComments = new HashMap<>();

        Pattern pattern = Pattern.compile("<!--(.*?)-->", Pattern.DOTALL);
        Matcher matcher = pattern.matcher(content);
        while (matcher.find()) {
            String comment = matcher.group(1).trim();
            htmlComments.put(matcher.start() + 1, url.isEmpty() ? comment : url + " " + comment);
        }

        return htmlComments;
    }

    public static Map<Integer, Map<Integer, String>> extractJsComments(String content) {
        Map<Integer, Map<Integer, String>> jsComments = new HashMap<>();
        Pattern scriptPattern = Pattern.compile("<script[^>]*>(.*?)</script>", Pattern.DOTALL | Pattern.CASE_INSENSITIVE);
        Matcher scriptMatcher = scriptPattern.matcher(content);

        while (scriptMatcher.find()) {
            Map<Integer, String> scriptComments = new HashMap<>();
            String scriptContent = scriptMatcher.group(1);
            Scanner scanner = new Scanner(scriptContent);

            int lineCounter = 1;
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                Pattern commentPattern = Pattern.compile("//(.*?)$|/\\*(.*?)\\*/");
                Matcher commentMatcher = commentPattern.matcher(line);
                while (commentMatcher.find()) {
                    String comment = commentMatcher.group(1) != null ? commentMatcher.group(1) : commentMatcher.group(2);
                    scriptComments.put(lineCounter, comment.trim());
                }
                lineCounter++;
            }
            jsComments.put(scriptMatcher.start() + "<script>".length(), scriptComments);
            scanner.close();
        }
        return jsComments;
    }

    public static Map<Integer, String> scanForProhibited(Map<Integer, String> htmlComments, Map<Integer, Map<Integer, String>> jsComments) {
        Map<Integer, String> prohibitedFindings = new HashMap<>();
        try {
            Scanner scanner = new Scanner(CommentScanner.class.getResourceAsStream("prohib.json"), "UTF-8");
            StringBuilder jsonBuilder = new StringBuilder();
            while (scanner.hasNextLine()) {
                jsonBuilder.append(scanner.nextLine());
            }
            scanner.close();
            String jsonContent = jsonBuilder.toString();
            // Parse JSON content
            // Assuming JSON is in the format: [{"prohib1": "value1"}, {"prohib2": "value2"}, ...]
            // Adjust this according to the actual format of your JSON
            // Here I'm assuming each entry is a Map with a single key-value pair
            String[] entries = jsonContent.split(",");
            for (String entry : entries) {
                String[] parts = entry.split(":");
                String prohibitedWord = parts[0].trim().replaceAll("[\"{}]", "");
                String value = parts[1].trim().replaceAll("[\"{}]", "");
                for (Map.Entry<Integer, String> htmlEntry : htmlComments.entrySet()) {
                    if (htmlEntry.getValue().toLowerCase().contains(value.toLowerCase())) {
                        prohibitedFindings.put(htmlEntry.getKey(), "PROHIBITED VALUE (" + prohibitedWord + ") in comment " + htmlEntry.getValue());
                    }
                }
                for (Map.Entry<Integer, Map<Integer, String>> scriptEntry : jsComments.entrySet()) {
                    for (Map.Entry<Integer, String> scriptCommentEntry : scriptEntry.getValue().entrySet()) {
                        if (scriptCommentEntry.getValue().toLowerCase().contains(value.toLowerCase())) {
                            prohibitedFindings.put(scriptEntry.getKey() + scriptCommentEntry.getKey() - 1, "PROHIBITED VALUE (" + prohibitedWord + ") in comment " + scriptCommentEntry.getValue());
                        }
                    }
                }
            }
        } catch (IOException e) {
            System.out.println("Error: Could not find prohib.json file");
        }
        return prohibitedFindings;
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java scan <url>");
            System.exit(1);
        }

        String url = args[0];
        try {
            Scanner scanner = new Scanner(new URL(url).openStream(), "UTF-8").useDelimiter("\\A");
            String content = scanner.hasNext() ? scanner.next() : "";
            scanner.close();

            Map<Integer, String> htmlComments = extractHtmlComments(content, url);
            for (Map.Entry<Integer, String> entry : htmlComments.entrySet()) {
                System.out.println("[" + entry.getKey() + "] " + entry.getValue());
            }

            Map<Integer, Map<Integer, String>> jsComments = extractJsComments(content);
            for (Map.Entry<Integer, Map<Integer, String>> scriptEntry : jsComments.entrySet()) {
                int scriptLocation = scriptEntry.getKey();
                Map<Integer, String> scriptComments = scriptEntry.getValue();
                for (Map.Entry<Integer, String> commentEntry : scriptComments.entrySet()) {
                    int line = commentEntry.getKey();
                    String comment = commentEntry.getValue();
                    System.out.println("[Script location: " + scriptLocation + ", Line: " + line + "] " + comment);
                }
            }

            Map<Integer, String> prohibitedFindings = scanForProhibited(htmlComments, jsComments);
            if (!prohibitedFindings.isEmpty()) {
                System.out.println("\n**************** PROHIBITED COMMENTS DISCOVERED ****************");
                for (Map.Entry<Integer, String> entry : prohibitedFindings.entrySet()) {
                    System.out.println(entry.getValue());
                }
            }
        } catch (IOException e) {
            System.out.println("Error: Could not retrieve content from URL: " + e.getMessage());
            System.exit(1);
        }
    }
}
